multishard_mutation_query.cc

```diff
@@ -781,6 +782,44 @@ future<foreign_ptr<lw_shared_ptr<typename ResultBuilder::result_type>>> do_query
     co_return f.get0();
 }
 
+template <typename ResultBuilder>
+future<foreign_ptr<lw_shared_ptr<typename ResultBuilder::result_type>>> do_query_tablets(
+        distributed<replica::database>& db,
+        schema_ptr s,
+        const query::read_command& cmd,
+        const dht::partition_range_vector& ranges,
+        tracing::trace_state_ptr trace_state,
+        db::timeout_clock::time_point timeout,
+        noncopyable_function<ResultBuilder()> result_builder_factory) {
+    auto& table = db.local().find_column_family(s);
+    auto erm = table.get_effective_replication_map();
+    const auto& token_metadata = erm->get_token_metadata();
+    const auto& tablets = token_metadata.tablets().get_tablet_map(s->id());
+    const auto this_node_id = token_metadata.get_topology().this_node()->host_id();
+
+    auto query_cmd = cmd;
+    auto result_builder = result_builder_factory();
+
+    std::vector<foreign_ptr<lw_shared_ptr<typename ResultBuilder::result_type>>> results;
+    locator::tablet_range_splitter range_splitter{s, tablets, this_node_id, ranges};
+    while (auto range_opt = range_splitter()) {
+        auto& r = *results.emplace_back(co_await db.invoke_on(range_opt->shard,
+                    [&result_builder, gs = global_schema_ptr(s), &query_cmd, &range_opt, gts = tracing::global_trace_state_ptr(trace_state), timeout] (replica::database& db) {
+            return result_builder.query(db, gs, query_cmd, range_opt->range, gts, timeout);
+        }));
+
+        // Substract result from limit, watch for underflow
+        query_cmd.partition_limit -= std::min(query_cmd.partition_limit, ResultBuilder::get_partition_count(r));
+        query_cmd.set_row_limit(query_cmd.get_row_limit() - std::min(query_cmd.get_row_limit(), ResultBuilder::get_row_count(r)));
+
+        if (!query_cmd.partition_limit || !query_cmd.get_row_limit() || r.is_short_read()) {
+            break;
```

Comment:
See? no append after a short read.

Reply:
Yes, I take care of ithere (not continuing merging after short read).

Reply:
This makes a stronger case for pushing iteration on tablets to the caller (or all the way up to the client). The most common result will be a short read (since tablets are 5GB and pages are 1MB), so most of the time the merging code does nothing.

But it's okay to get there later.